libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/EnhancedLocationEngineDecorator.java

```diff
@@ -0,0 +1,221 @@
+import com.mapbox.android.core.location.LocationEngineCallback;
+import com.mapbox.android.core.location.LocationEngineRequest;
+import com.mapbox.android.core.location.LocationEngineResult;
+import com.mapbox.geojson.Point;
+import com.mapbox.navigator.FixLocation;
+import com.mapbox.navigator.NavigationStatus;
+import com.mapbox.services.android.navigation.v5.internal.navigation.MapboxNavigator;
+
+import java.io.File;
+import java.lang.ref.WeakReference;
+import java.util.Date;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import timber.log.Timber;
+
+class EnhancedLocationEngineDecorator implements LocationEngine {
+  private static final long UPDATE_INTERVAL_IN_MILLISECONDS = 1000;
+  private static final long FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS = 500;
+  private final LocationEngine locationEngine;
+  private final MapboxNavigator mapboxNavigator;
+  private final LocationEngineCallback<LocationEngineResult> callback = new CurrentLocationEngineCallback(this);
+  private ScheduledFuture future;
+  private final ScheduledExecutorService executorService;
+  private Location rawLocation;
+  private LocationEngineCallback<LocationEngineResult> originalCallback;
+  private Looper originalLooper;
+  private boolean isEnhancedInitialized = false;
+  private AtomicBoolean isConfigured = new AtomicBoolean(false);
+
+  EnhancedLocationEngineDecorator(@NonNull LocationEngine locationEngine,
+                                  @NonNull MapboxNavigator mapboxNavigator,
+                                  @NonNull ScheduledExecutorService executorService) {
+    this.locationEngine = locationEngine;
+    this.mapboxNavigator = mapboxNavigator;
+    this.executorService = executorService;
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    locationEngine.getLastLocation(callback);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull final LocationEngineCallback<LocationEngineResult> callback,
+                                     @Nullable Looper looper) throws SecurityException {
+    if (!isEnhancedInitialized) {
+      isEnhancedInitialized = true;
+      originalCallback = callback;
+      originalLooper = looper;
+      onFreeDrive();
+    } else {
+      locationEngine.requestLocationUpdates(request, callback, looper);
+    }
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     PendingIntent pendingIntent) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, pendingIntent);
+  }
+
+  @Override
+  public void removeLocationUpdates(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
```

Comment:
This should remove the original callback+looper pair from the map/list and stop listening with our internal callback (in free drive mode) if the map/list is now empty. If the callback is not in our list, it should wire to the original engine.